solr/core/src/java/org/apache/solr/cloud/WatchedClusterProperties.java

```diff
@@ -0,0 +1,113 @@
+package org.apache.solr.cloud;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.concurrent.CopyOnWriteArrayList;
+import java.util.function.BiConsumer;
+import org.apache.solr.common.cloud.ClusterPropertiesListener;
+
+public class WatchedClusterProperties implements ClusterPropertiesListener {
+
+  final String nodeName;
+
+  public static final String WATCHED_PROPERTIES = "watched-properties";
+  public static final String WATCHED_NODE_PROPERTIES = "watched-node-properties";
+
+  private Map<String, List<BiConsumer<String, String>>> propertiesListeners =
+      Collections.synchronizedMap(new HashMap<>());
+
+  private Map<String, List<BiConsumer<String, String>>> nodePropertiesListeners =
+      Collections.synchronizedMap(new HashMap<>());
+
+  private volatile Map<String, String> knownData = new HashMap<>();
+  private volatile Map<String, String> knownNodeData = new HashMap<>();
+
+  public WatchedClusterProperties(String nodeName) {
+    this.nodeName = nodeName;
+  }
+
+  @Override
+  @SuppressWarnings("unchecked")
+  public boolean onChange(Map<String, Object> properties) {
+    Map<String, String> newData =
+        (Map<String, String>) properties.getOrDefault(WATCHED_PROPERTIES, Collections.EMPTY_MAP);
+    Map<String, String> modified =
+        compareAndInvokeListeners(newData, knownData, propertiesListeners);
+    if (modified != null) knownData = modified;
+
+    newData =
+        ((Map<String, Map<String, String>>)
+                properties.getOrDefault(WATCHED_NODE_PROPERTIES, Collections.EMPTY_MAP))
+            .getOrDefault(nodeName, Collections.EMPTY_MAP);
+    modified = compareAndInvokeListeners(newData, knownNodeData, nodePropertiesListeners);
+    if (modified != null) knownNodeData = modified;
+    return false;
+  }
+
+  private static Map<String, String> compareAndInvokeListeners(
+      Map<String, String> newData,
+      Map<String, String> knownData,
+      Map<String, List<BiConsumer<String, String>>> propertiesListeners) {
+    boolean isModified = false;
+    // look for changed data or missing keys
+    for (String k : knownData.keySet()) {
+      String oldVal = knownData.get(k);
+      String newVal = newData.get(k);
+      if (Objects.equals(oldVal, newVal)) continue;
+      isModified = true;
+      invokeListener(k, newVal, propertiesListeners);
+    }
+    for (String k : newData.keySet()) {
+      if (knownData.containsKey(k)) continue;
+      isModified = true;
+      invokeListener(k, newData.get(k), propertiesListeners);
```

Comment:
yeah, that is what it does. You register the watcher from within anode and listeners are fired for the properties of that node only

Reply:
What will happens we update the property in ZK, will that cause listener-update-event on each node or all node?

Reply:
what do you mean by each node and all node?

everyone who registered a listener should get a callback

Reply:
For example I want these property value for this node only
```
{
  component: "basiclimit", host:solr-1,property-x:value-y
} 
```
Then this should be invoked on host solr-1 only.

Reply:
when the  property is defined for ```solr-1```  , you only get a calback on ```solr-1``` only
It would look something like this

```
{
  "watched-node-properties" :{
    "solr-1" : {
        "basiclimit.property-x":"value-y"
    }
  }
}

```

Reply:
Nice, let's give demo on this in Monday's sync. thanks!